persistence/sql/persister_identity.go

```diff
@@ -249,37 +250,67 @@ func (p *Persister) CreateIdentity(ctx context.Context, i *identity.Identity) er
 	})
 }
 
-func (p *Persister) ListIdentities(ctx context.Context, page, perPage int) ([]identity.Identity, error) {
-	is := make([]identity.Identity, 0)
-
-	/* #nosec G201 TableName is static */
-	if err := sqlcon.HandleError(p.GetConnection(ctx).Where("nid = ?", corp.ContextualizeNID(ctx, p.nid)).
-		EagerPreload("VerifiableAddresses", "RecoveryAddresses").
-		Paginate(page, perPage).Order("id DESC").
-		All(&is)); err != nil {
-		return nil, err
-	}
-
-	schemaCache := map[string]string{}
+func (p *Persister) setIdentitySchemas(ctx context.Context, is []identity.Identity) error {
+	var schemaCache = make(map[string]string)
 
 	for k := range is {
 		i := &is[k]
 		if err := i.ValidateNID(); err != nil {
-			return nil, sqlcon.HandleError(err)
+			return sqlcon.HandleError(err)
 		}
 
 		if u, ok := schemaCache[i.SchemaID]; ok {
 			i.SchemaURL = u
 		} else {
 			if err := p.injectTraitsSchemaURL(ctx, i); err != nil {
-				return nil, err
+				return err
 			}
 			schemaCache[i.SchemaID] = i.SchemaURL
 		}
 
 		is[k] = *i
 	}
 
+	return nil
+}
+
+func (p *Persister) ListIdentities(ctx context.Context, page, perPage int) ([]identity.Identity, error) {
+	is := []identity.Identity{}
+
+	/* #nosec G201 TableName is static */
+	if err := sqlcon.HandleError(p.GetConnection(ctx).Where("nid = ?", corp.ContextualizeNID(ctx, p.nid)).
+		EagerPreload("VerifiableAddresses", "RecoveryAddresses").
+		Paginate(page, perPage).Order("id DESC").
+		All(&is)); err != nil {
+		return nil, err
+	}
+
+	if err := p.setIdentitySchemas(ctx, is); err != nil {
+		return nil, err
+	}
+
+	return is, nil
+}
+
+func (p *Persister) ListIdentitiesFiltered(ctx context.Context, filters map[string][]string, page, perPage int) ([]identity.Identity, error) {
```

Comment:
This function needs tests ensure that filters work and no SQL injections can happen.

Reply:
@david972 this function needs several tests (you can write table tests)! The tests for this are located here: https://github.com/ory/kratos/blob/master/identity/test/pool.go